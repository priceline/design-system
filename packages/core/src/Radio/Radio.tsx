import { RadioChecked, RadioEmpty } from 'pcln-icons'
import React, { type ComponentPropsWithoutRef } from 'react'
import styled from 'styled-components'
import { type PaletteFamilyName } from '../theme/theme'
import { applyVariations, getPaletteColor } from '../utils/utils'

const RadioWrap = styled.div`
  display: inline-block;
  color: ${getPaletteColor('text.light')};
  cursor: ${(props) => (props.disabled ? `not-allowed` : `pointer`)};
  &:hover > svg {
    ${(props) => {
      if (props.checked && !props.disabled) {
        return `color: ${getPaletteColor('dark')(props)} !important;`
      }

      return props.disabled ? null : `color: ${getPaletteColor('base')(props)};`
    }};
  }
  ${applyVariations('Radio')}
`

const RadioInput = styled.input`
  appearance: none;
  opacity: 0;
  position: absolute;
  z-index: 0;

  & ~ svg {
    border: 1px solid transparent;
    border-radius: 50%;
    padding: 2px;
  }

  &:focus {
    box-shadow: none;
    & ~ svg {
      border: 1px solid ${getPaletteColor('border.base')};
      background-color: ${getPaletteColor('background.light')};
    }
  }
  &:checked ~ svg {
    color: ${getPaletteColor('base')};
  }
  &:disabled ~ svg {
    color: ${getPaletteColor('border.base')};
    cursor: not-allowed;
  }
`

const RadioCheckedIcon = styled(RadioChecked)`
  vertical-align: middle;
`
const RadioEmptyIcon = styled(RadioEmpty)`
  vertical-align: middle;
`

type RadioIconProps = {
  checked?: boolean
  size?: number
}

const RadioIcon = ({ checked, ...props }: RadioIconProps) => {
  return checked ? <RadioCheckedIcon {...props} /> : <RadioEmptyIcon {...props} />
}

/**
 * @public
 */
export type RadioProps = ComponentPropsWithoutRef<'input'> & {
  size?: number
  color?: PaletteFamilyName
}

/**
 * @public
 */
export const Radio = React.forwardRef<HTMLInputElement, RadioProps>(
  ({ color = 'primary', size = 24, ...props }, ref) => {
    const defaultedProps = { color, size, ...props }
    const { checked, disabled } = defaultedProps

    const borderAdjustedSize = size + 4

    const dataNameHelper = (checked, disabled) => {
      if ((checked && disabled) || (!checked && disabled)) {
        return 'disabled'
      }
      if (checked && !disabled) {
        return 'checked'
      }
      return 'unchecked'
    }

    return (
      <RadioWrap
        color={color}
        checked={checked}
        disabled={disabled}
        data-name={dataNameHelper(checked, disabled)}
      >
        <RadioInput type='radio' {...defaultedProps} ref={ref} />
        <RadioIcon checked={checked} size={borderAdjustedSize} />
      </RadioWrap>
    )
  }
)

Radio.displayName = 'Radio'
