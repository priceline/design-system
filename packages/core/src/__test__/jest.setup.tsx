import '@testing-library/jest-dom'
import 'jest-styled-components'

global.requestAnimationFrame =
  global.requestAnimationFrame ||
  function _raf(cb) {
    return setTimeout(cb, 0)
  }
