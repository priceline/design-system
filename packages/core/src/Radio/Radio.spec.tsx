import { render } from '../__test__/testing-library'
import { Radio } from './Radio'

describe('Radio', () => {
  test('Selected, rendering', () => {
    const { asFragment } = render(<Radio checked />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('Not Selected, rendering', () => {
    const { asFragment } = render(<Radio />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('Disabled, rendering', () => {
    const { asFragment } = render(<Radio disabled />)
    expect(asFragment()).toMatchSnapshot()
  })
})
