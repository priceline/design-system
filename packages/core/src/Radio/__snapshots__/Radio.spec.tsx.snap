// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio Disabled, rendering 1`] = `
<DocumentFragment>
  .c0 {
  font-family: 'Montserrat','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.4;
  font-weight: 500;
}

.c0 * {
  box-sizing: border-box;
}

.c3 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  line-height: 1;
  width: 28px;
}

.c4 {
  outline: none;
}

.c1 {
  display: inline-block;
  color: #4f6f8f;
  cursor: not-allowed;
}

.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  opacity: 0;
  position: absolute;
  z-index: 0;
}

.c2 ~ svg {
  border: 1px solid transparent;
  border-radius: 50%;
  padding: 2px;
}

.c2:focus {
  box-shadow: none;
}

.c2:focus ~ svg {
  border: 1px solid #c0cad5;
  background-color: #f4f6f8;
}

.c2:checked ~ svg {
  color: #0068ef;
}

.c2:disabled ~ svg {
  color: #c0cad5;
  cursor: not-allowed;
}

.c5 {
  vertical-align: middle;
}

<div
    class="c0"
  >
    <div
      class="c1"
      color="primary"
      data-name="disabled"
      disabled=""
    >
      <input
        class="c2"
        color="primary"
        disabled=""
        size="24"
        type="radio"
      />
      <svg
        aria-hidden="true"
        class="c3 c4 c5"
        fill="currentcolor"
        focusable="false"
        height="28"
        role="img"
        size="28"
        tabindex="-1"
        viewBox="0 0 24 24"
        width="28"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 2C6.5 2 2 6.5 2 12s4.5 10 10 10 10-4.5 10-10S17.5 2 12 2zm0 18c-4.4 0-8-3.6-8-8s3.6-8 8-8 8 3.6 8 8-3.6 8-8 8z"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Radio Not Selected, rendering 1`] = `
<DocumentFragment>
  .c0 {
  font-family: 'Montserrat','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.4;
  font-weight: 500;
}

.c0 * {
  box-sizing: border-box;
}

.c3 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  line-height: 1;
  width: 28px;
}

.c4 {
  outline: none;
}

.c1 {
  display: inline-block;
  color: #4f6f8f;
  cursor: pointer;
}

.c1:hover > svg {
  color: #0068ef;
}

.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  opacity: 0;
  position: absolute;
  z-index: 0;
}

.c2 ~ svg {
  border: 1px solid transparent;
  border-radius: 50%;
  padding: 2px;
}

.c2:focus {
  box-shadow: none;
}

.c2:focus ~ svg {
  border: 1px solid #c0cad5;
  background-color: #f4f6f8;
}

.c2:checked ~ svg {
  color: #0068ef;
}

.c2:disabled ~ svg {
  color: #c0cad5;
  cursor: not-allowed;
}

.c5 {
  vertical-align: middle;
}

<div
    class="c0"
  >
    <div
      class="c1"
      color="primary"
      data-name="unchecked"
    >
      <input
        class="c2"
        color="primary"
        size="24"
        type="radio"
      />
      <svg
        aria-hidden="true"
        class="c3 c4 c5"
        fill="currentcolor"
        focusable="false"
        height="28"
        role="img"
        size="28"
        tabindex="-1"
        viewBox="0 0 24 24"
        width="28"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 2C6.5 2 2 6.5 2 12s4.5 10 10 10 10-4.5 10-10S17.5 2 12 2zm0 18c-4.4 0-8-3.6-8-8s3.6-8 8-8 8 3.6 8 8-3.6 8-8 8z"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Radio Selected, rendering 1`] = `
<DocumentFragment>
  .c0 {
  font-family: 'Montserrat','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.4;
  font-weight: 500;
}

.c0 * {
  box-sizing: border-box;
}

.c3 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  line-height: 1;
  width: 28px;
}

.c4 {
  outline: none;
}

.c1 {
  display: inline-block;
  color: #4f6f8f;
  cursor: pointer;
}

.c1:hover > svg {
  color: #049 !important;
}

.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  opacity: 0;
  position: absolute;
  z-index: 0;
}

.c2 ~ svg {
  border: 1px solid transparent;
  border-radius: 50%;
  padding: 2px;
}

.c2:focus {
  box-shadow: none;
}

.c2:focus ~ svg {
  border: 1px solid #c0cad5;
  background-color: #f4f6f8;
}

.c2:checked ~ svg {
  color: #0068ef;
}

.c2:disabled ~ svg {
  color: #c0cad5;
  cursor: not-allowed;
}

.c5 {
  vertical-align: middle;
}

<div
    class="c0"
  >
    <div
      class="c1"
      color="primary"
      data-name="checked"
    >
      <input
        checked=""
        class="c2"
        color="primary"
        size="24"
        type="radio"
      />
      <svg
        aria-hidden="true"
        class="c3 c4 c5"
        fill="currentcolor"
        focusable="false"
        height="28"
        role="img"
        size="28"
        tabindex="-1"
        viewBox="0 0 24 24"
        width="28"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 7c-2.8 0-5 2.2-5 5s2.2 5 5 5 5-2.2 5-5-2.2-5-5-5zm0-5C6.5 2 2 6.5 2 12s4.5 10 10 10 10-4.5 10-10S17.5 2 12 2zm0 18c-4.4 0-8-3.6-8-8s3.6-8 8-8 8 3.6 8 8-3.6 8-8 8z"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
